### PR TITLE
feat(scores): add annotation slider and score config settings for categorical annotation

### PR DESCRIPTION
**Pull request by marliessophie**

### CODE REVIEW

File: packages/shared/prisma/seed.ts
Comment (marliessophie) [inline packages/shared/prisma/seed.ts:1]: Will add some more helpful data to seeder, still left on to-do list 👍

Comment (marliessophie) [inline packages/shared/prisma/seed.ts:1]: FYI only Seeder data will still be improved before launch

File: pnpm-lock.yaml
Comment (marliessophie) [inline pnpm-lock.yaml:1]: See dependency installation question above, thanks

File: package.json [hunk 1]
```diff
@@ -69,5 +69,9 @@
         "pr": ":rocket: _This pull request is included in v${version}. See [${releaseName}](${releaseUrl}) for release notes._"
       }
     }
+  },
+  "dependencies": {
+    "@radix-ui/react-toggle-group": "^1.0.4",
+    "vaul": "^0.9.1"
```
[the remark this answers is not in the document]
Comment (marcklingen) [reply package.json:75]: you need to cd into the web proejct and install dependencies there. that's why you probably got an alert by pnpm to check whether you really want to do an install on the workspace level

File: pnpm-lock.yaml
Comment (marcklingen) [reply pnpm-lock.yaml:1]: yep, need to move these to the `web` project

File: package.json (hunk 1 above)
[the remark this answers is not in the document]
Comment (marliessophie) [reply package.json:75]: thanks, will do 🙏

File: packages/shared/prisma/migrations/20240531155458_backfill_manual_scores/migration.sql
Comment (marliessophie) [inline packages/shared/prisma/migrations/20240531155458_backfill_manual_scores/migration.sql:1]: Would be good to review this migration please

File: web/src/features/manual-scoring/components/multi-select-key-values.tsx
Comment (marliessophie) [inline web/src/features/manual-scoring/components/multi-select-key-values.tsx:1]: I created this component in a generic fashion to be able to replace the mutli-select component used in filters to only use this one. It should work out of the box but did not want to pollute this PR with an unrelated change. Have added to backlog.

**Review by marcklingen — COMMENTED** [review]
backfill notes

File: web/src/features/manual-scoring/components/multi-select-key-values.tsx
Comment (marcklingen) [reply web/src/features/manual-scoring/components/multi-select-key-values.tsx:1]: looks really nice

File: web/src/server/api/routers/scores.ts [hunk 2]
```diff
@@ -149,38 +159,49 @@ export const scoresRouter = createTRPCRouter({
         throw new Error("No trace with this id in this project.");
       }
 
-      const score = await ctx.prisma.score.create({
-        data: {
+      try {
+        const score = await ctx.prisma.score.create({
```
Comment (marcklingen) [inline web/src/server/api/routers/scores.ts:163]: use upsert to make sure that no score config can be applied twice via annotation workflow

File: web/src/components/trace/index.tsx [hunk 3]
```diff
@@ -103,11 +104,13 @@ export function Trace(props: {
             trace={props.trace}
             observations={props.observations}
             scores={props.scores}
+            configs={props.configs}
```
Comment (marcklingen) [inline web/src/components/trace/index.tsx:107]: we do not need to pass the prop down, we could just fetch the config down in the annotation drawer as react query will cache it locally if is needed mutliple times within the same session

File: web/src/features/manual-scoring/components/multi-select-key-values.tsx [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+import { Separator } from "@/src/components/ui/separator";
+
+type MultiSelectOptions = {
+  value: string;
+  key?: string;
+  count?: number;
+  disabled?: boolean;
+};
+
+export function MultiSelectKeyValues<
```
Comment (marcklingen) [inline web/src/features/manual-scoring/components/multi-select-key-values.tsx:30]: move to generic components (src/components)?

File: web/src/server/api/routers/scoreConfigs.ts [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+  projectId: z.string(), // Required for protectedProjectProcedure
+});
+
+const category = z.object({
+  label: z.string().min(1),
+  value: z.number(),
+});
+
+export const scoreConfigsRouter = createTRPCRouter({
+  all: protectedProjectProcedure
```
Comment (marcklingen) [inline web/src/server/api/routers/scoreConfigs.ts:20]: add pagination

File: web/src/server/api/routers/scoreConfigs.ts [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+        return {
+          configs,
+          totalCount: configsCount,
+        };
+      } catch (error) {
+        console.log(error);
+      }
+    }),
+
+  create: protectedProjectProcedure
```
Comment (marcklingen) [inline web/src/server/api/routers/scoreConfigs.ts:54]: check if no other config with the same name exists?

File: web/src/features/manual-scoring/components/multi-select-key-values.tsx (hunk 4 above)
Comment (marliessophie) [reply web/src/features/manual-scoring/components/multi-select-key-values.tsx:30]: Agree! Was planning to do in follow up, does that work?

Comment (marcklingen) [reply web/src/features/manual-scoring/components/multi-select-key-values.tsx:30]: follow up for marlies

File: web/src/server/api/routers/scoreConfigs.ts (hunk 5 above)
Comment (marcklingen) [reply web/src/server/api/routers/scoreConfigs.ts:20]: future

File: packages/shared/prisma/migrations/20240604133340_backfill_manual_scores/migration.sql [hunk 7]
```diff
@@ -0,0 +1,50 @@
+-- Create manual-score config for any project with manual scores and link config_id to scores
+BEGIN;
+WITH project_configs AS (
```
Comment (maxdeichmann) [inline packages/shared/prisma/migrations/20240604133340_backfill_manual_scores/migration.sql:3]: I think we could have found a simpler way to move this data, which might be less elegant and produces some garbage data but cleans up. As this would happen in a transaction, this would be save anyways.

File: packages/shared/prisma/migrations/20240604133340_backfill_manual_scores/migration.sql [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+			sc.name = 'manual-score'
+			AND sc.project_id = s.project_id)
+	RETURNING
+		id,
+		project_id
+)
+UPDATE
+	scores
+SET
+	config_id = pc.id
```
Comment (maxdeichmann) [inline packages/shared/prisma/migrations/20240604133340_backfill_manual_scores/migration.sql:44]: i assume all scores with 'manual-score' in the same project should have the same config_id?

Comment (marliessophie) [reply packages/shared/prisma/migrations/20240604133340_backfill_manual_scores/migration.sql:44]: Correct 👍
